### PR TITLE
`ModelingLLM`: Add Structured Grading Instruction Generation and Restructure Module

### PR DESCRIPTION
**Pull request by LeonWehrhahn**
### Motivation and Context

Structured Grading Instructions (SGIs) lead to more accurate feedback generations. However, instructors do not always provide SGIs for every exercise. To address this gap and maintain consistency in the grading process, this PR implements automatic SGI generation when manual ones are not present.

Additionally, as the module has grown with more functions for processing and manipulating data, this PR includes a restructuring to enhance organization and maintainability.

### Description

> A follow-up PR will address caching the structured grading instructions

This PR introduces the following key changes:

1. **Automatic SGI Generation:** Implemented a new function to generate Structured Grading Instructions when they are not manually provided for an exercise.
2. **Module Restructuring:** Reorganized the module structure to better accommodate the growing number of processing and manipulation functions.
3. **Core Functionality Updates:**
    - Restructured generate_suggestions.py to focus solely on the feedback generation process and moved other functionalities into separate functions and files for improved modularity.
    - Add get_structured_grading_instructions.py to include the new SGI auto-generation feature
4. **Model and Utility Improvements:**
   - Added new models for Structured Grading Instructions.
   - Added new utility functions to support the core functionality changes
5. **Prompt Updates:** Refined existing prompts and added new ones to support the automatic SGI generation

### Steps for Testing

1. Make direct API calls to the /feedback_suggestions endpoint of the module.
2. Test with various diagram types (BPMN, UML Class, UML Activity, etc.) to ensure the functionality works for all Apollon-supported diagrams.
3. Test both scenarios where Structured Grading Instructions (SGIs) are manually provided and where they are not.
4. Verify that appropriate feedback is generated in both cases.

[](https://athena-test1.ase.cit.tum.de)
[](https://athena-test2.ase.cit.tum.de)

### CODE REVIEW

**Review by FelixTJDietrich — CHANGES_REQUESTED** [review]
Please resolve the conflicts

File: modules/modeling/module_modeling_llm/module_modeling_llm/apollon_transformer/parser/element.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
+from typing import Dict, Any, List, Optional
+
+class Element:
```
Comment (undernagruzez) [inline modules/modeling/module_modeling_llm/module_modeling_llm/apollon_transformer/parser/element.py]: Is it possible to give the class a more specific name? Element seems to be too general.

File: modules/modeling/module_modeling_llm/module_modeling_llm/apollon_transformer/parser/element.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    """
+    Represents an element in a UML diagram.
+
```
Comment (undernagruzez) [inline modules/modeling/module_modeling_llm/module_modeling_llm/apollon_transformer/parser/element.py]:
```suggestion
```

**Review by undernagruzez — COMMENTED** [review]
Two small suggestions
Also a question: why do we need the parser scripts in this PR? Weren't they already merged in the previous PR?

File: playground/package-lock.json
Comment (maximiliansoelch) [inline playground/package-lock.json:1]: Should this change be part of this PR?

File: playground/.gitignore
Comment (maximiliansoelch) [inline playground/.gitignore:1]: Should this change be part of this PR?

File: modules/modeling/module_modeling_llm/module_modeling_llm/utils/predict_and_parse.py [hunk 3]
```diff
@@ -15,18 +14,6 @@ async def predict_and_parse(
...
-    Args:
-        model (BaseLanguageModel): The model to predict with
-        chat_prompt (ChatPromptTemplate): Prompt to use
-        prompt_input (dict): Input parameters to use for the prompt
-        pydantic_object (Type[T]): Pydantic model to parse the output
-        tags (Optional[List[str]]: List of tags to tag the prediction with
-
-    Returns:
-        Optional[T]: Parsed output, or None if it could not be parsed
-    """
```
Comment (maximiliansoelch) [inline modules/modeling/module_modeling_llm/module_modeling_llm/utils/predict_and_parse.py:29]: Why is this comment outdated?

**Review by maximiliansoelch — CHANGES_REQUESTED** [review]

File: modules/modeling/module_modeling_llm/module_modeling_llm/utils/predict_and_parse.py (hunk 3 above)
Comment (maximiliansoelch) [reply modules/modeling/module_modeling_llm/module_modeling_llm/utils/predict_and_parse.py:29]: Will be added in a follow up again

**Review by maximiliansoelch — APPROVED** [review]

**Pull request merged**